### PR TITLE
bpr fixes

### PR DESCRIPTION
**Pull request by andland**
In line 38, that would have been treated as (1:100) * row_x, which I don't think you want.

For the probability and the loss, I think there is some confusion. I re-derived the gradient and this is what I get (which is different than in the BPR paper). I am pretty sure there is an error in the BPR paper.

### CODE REVIEW

File: R/ALG_BPR.R [hunk 1]
```diff
@@ -35,7 +35,7 @@ BPR <- function(data, k = 10, lambda = 0.05, regU = 0.0025, regI = 0.0025, regJ
     
     while (!isConverged(x, p)) {
         
-        for (s in 1:100 * row_x) {
+        for (s in 1:(100 * row_x)) {
```
Comment (ludovikcoba) [inline R/ALG_BPR.R]: You are right about this!

Comment (andland) [inline R/ALG_BPR.R]: This python implementation calculates the updates the same way that I proposed: [https://github.com/gamboviol/bpr/blob/master/bpr.py#L80-L98](https://github.com/gamboviol/bpr/blob/master/bpr.py#L80-L98). `z = 1.0/(1.0+exp(x))`.

**Comment by andland** [comment]
Also, U and V should be randomly initialized. I ran the algorithm with the current initialization, and the learned vectors are all proportional to the 1 vector.

**Comment by ludovikcoba** [comment]
Thank you so much for the help! It took me some time to answer because I was traveling. About the random initialization I will get it ready the next days.

**Comment by ludovikcoba** [comment]
Hi Andrew! Would you mind to point me a paper for the BPR and to consider for the proper initialization of the features matrix.

**Comment by andland** [comment]
I don't have a reference, but it is necessary for some algorithms to be randomly initialized. The amount or type of initialization can affect the results. In libFM, Rendle defaults to initializing with i.i.d. normal with a standard deviation of 0.1. I have seen other use random uniform between -a and +a.

```
U <- matrix(rnorm(row_x * row_y, 0, 0.1), nrow = row_x, ncol = k)
```

**Comment by ludovikcoba** [comment]
Just to let you know, I am performing some more updates on the library and I will release it very soon, as well with the update on BPR.

**Pull request merged**